pika/channel.py

```diff
@@ -1267,9 +1269,40 @@ def _on_synchronous_complete(self, _method_frame_unused):
         """
         LOGGER.debug('%i blocked frames', len(self._blocked))
         self._blocking = None
+        # self._blocking must be checked here as a callback could
+        # potentially change the state of that variable during an
+        # iteration of the while loop
         while self._blocked and self._blocking is None:
             self._rpc(*self._blocked.popleft())
 
+    def _drain_blocked_methods_on_remote_close(self):
+        """This is called when the broker sends a Channel.Close while the
+        client is in CLOSING state. This method checks the blocked method
+        queue for a pending client-initiated Channel.Close method and
+        ensures its callbacks are processed, but does send the method to
+        the broker.
+
+        The broker may close the channel before responding to outstanding
+        (blocked) synchronous methods, or even before these methods have
+        been sent to the broker. AMQP 0.9.1 obliges the server to drop all
+        methods arriving on a closed channel other than Channel.CloseOk and
+        Channel.Close. Since the response to the blocked synchronous
+        method never arrives, the channel never becomes unblocked, and the
+        Channel.Close, if any, in the blocked queue has no opportunity to
+        be sent, and thus its completion callback would never be called.
```

Comment:
Corrections:

        The broker may close the channel before responding to outstanding in-transit
        synchronous methods, or even before these methods have
        been sent to the broker. AMQP 0.9.1 obliges the server to drop all
        methods arriving on a closed channel other than Channel.CloseOk and
        Channel.Close. Since the response to a synchronous method that blocked
        the channel never arrives, the channel never becomes unblocked, and the
        Channel.Close, if any, in the blocked queue has no opportunity to
        be sent, and thus its completion callback would never be called.

Reply:
@lukebakken - we're almost there. Just comment changes remain.